### PR TITLE
Use pybind11 in ttconv module

### PR DESCRIPTION
**Pull request by thomasjpfan**
### PR Summary
Towards #23846

This PR migrates the ttconv module to using pybind11. Most of the conversion uses `pybind11` data structures. Some notes for reviewers:

1. The Python C-API was required to converting bytes into Unicode: [pybind11 reference](https://pybind11.readthedocs.io/en/stable/advanced/cast/strings.html#explicit-conversions). 
2. [pybind11::handle](https://pybind11.readthedocs.io/en/stable/reference.html#_CPPv46handle) is a thin wrapper around `PyObject *` that does not perform any reference counting.

### CODE REVIEW

File: src/_ttconv.cpp
Comment (oscargus) [inline src/_ttconv.cpp:34]: Thank you very much for this! There are more suitable reviewers (@ianthomas23 and @anntzer comes to mind), but I think it would be good to replace this exception with "something else" (to get rid of the `py_exceptions.h` import, which we want to get rid of as part of the PyBind11 conversion). I *think* that `std::exception()` is the way to go, possibly with some sensible error message. Although, of course, that general exception is not good etc, but at least it will be consistent with the current behavior.

(This exception is not tested in the original code either.)

Comment (anntzer) [reply src/_ttconv.cpp:34]: This should be py::error_already_set (https://pybind11.readthedocs.io/en/stable/advanced/exceptions.html#handling-exceptions-from-python-in-c).  We should probably switch the whole thing to write to binary streams (and do something like fh.flush(); convert_ttf_to_ps(..., fh.buffer, ...); fh.buffer.flush() on the Python calling side), but that can be another PR.

Comment (thomasjpfan) [reply src/_ttconv.cpp]: I switched to using `error_already_set`. Note that this PR is keeping the original behavior on `main` which catches this exception and ignores it:

https://github.com/matplotlib/matplotlib/blob/de8e9d01b0d5bb19c9916f1b49b7eb207899b09a/src/_ttconv.cpp#L139-L142

Comment (anntzer) [reply src/_ttconv.cpp]: > catches this exception and ignores it
No, returning NULL is the way for C-API functions to signal that an exception has been set and should be propagated to python-land...

Comment (thomasjpfan) [reply src/_ttconv.cpp]: Ah I see. I misinterpreted the original implementation. Thank you for the explanation!

**Review by anntzer — APPROVED** [review]

File: src/_ttconv.cpp
Comment (anntzer) [inline src/_ttconv.cpp:77]: default value for glyph_ids is missing.

File: src/_ttconv.cpp
Comment (anntzer) [inline src/_ttconv.cpp:41]: Just make this argument a std::vector<int> (probably a `std::optional<std::vector<int>>`, in fact) and don't bother with the conversion?

**Comment by anntzer** [comment]
Frankly, I would not bother with changing ttconv; I'd rather just see some work on reviving #20866 to move everything to fonttools (which we already did for pdf in #20391).

**Comment by oscargus** [comment]
> I would not bother with changing ttconv;

Although I agree in principle, the question is if/when that will happen... If it is before 3.8 is released (i.e., in July(?)), the current PR may not be worthwhile, but considering that it has been around for 14 months it is not obvious that it will happen.

Anyway, good that you bring it up, so your call @thomasjpfan . We appreciate your work, but there is a risk that it will be in vain in the sense that it will be removed before the next release. On the other hand, I still think it is worthwhile to merge this if finished as one do not know if it will be replaced and there are only details remaining. Sorry we didn't make it clear in #23846 that for this particular module there was another possible route.

File: src/_ttconv.cpp
Comment (thomasjpfan) [reply src/_ttconv.cpp]: With the default value of `None`, I think this needs to be `pybind11::iterable*`, to account for the `nullptr`. If we need to use `std::vector<int>`, then we need `std::optional`, which requires `C++17`:

XREF: The note section from: https://pybind11.readthedocs.io/en/stable/advanced/functions.html#allow-prohibiting-none-arguments

**Comment by thomasjpfan** [comment]
> but there is a risk that it will be in vain in the sense that it will be removed before the next release.

I am okay with the changes made in this PR being removed because of #20866.

> On the other hand, I still think it is worthwhile to merge

I agree that it makes sense to move forward with this PR anyways to get the benefits of using `pybind11` as stated in https://github.com/matplotlib/matplotlib/issues/23846.

File: src/_ttconv.cpp
Comment (ianthomas23) [inline src/_ttconv.cpp:52]: It is really petty of me, but these two lines of whitespace can just be one.

File: src/_ttconv.cpp
Comment (ianthomas23) [inline src/_ttconv.cpp:64]: I am concerned that we don't exercise this line in the test suite as it would be good to confirm that `TTException`s propagate through `pybind11` as expected. But given that we've evidently never had such a test it doesn't have to be done here, it can be punted to "future work".

File: src/_ttconv.cpp
Comment (ianthomas23) [inline src/_ttconv.cpp:68]: This catch all isn't really needed as `pybind11` catches C++ exceptions and converts them to Python ones. But I like the approach that support for `pybind11` should happen with minimal changes to C++ code so it is fine to leave it as it is.

**Review by ianthomas23 — APPROVED** [review]
I've left a few comments but I'd be happy for this to merged as it is.

**Comment by oscargus** [comment]
I fixed the merge conflict (in the browser...). Is there anything left here @anntzer? (Now, assuming that we are not fully sure that #20866 will be implemented in time for 3.8 and that we want to swap to PyBind11 "ASAP".)

**Comment by anntzer** [comment]
From a quick look it seems fine.  Let's just get this in modulo CI.

**Review by anntzer — APPROVED** [review]

**Comment by oscargus** [comment]
Thanks @thomasjpfan !

**Pull request merged**
